### DISCUSSION TITLE
What's the most efficient way to test for multiple volume averages?

### DISCUSSION THREAD

**Discussion by contenrico** [casact/chainladder-python#465] [General]
Hello,

I'm running many scenarios on many reserving classes to test sensitivities. These scenarios are mostly varying the averaging period, so in the context of the Development method, I'm looping in a similar way:

`drop = my_long_list`
`for i in range(2, 12):`
`    cl.Development(n_periods=i, drop=my_long_list).fit(triangle).ldf_`

The drop argument always takes the same input, but because it's a very long list of exclusions, every iteration takes over a second to run, so running a lot of classes becomes quite slow.

Is there a way to fit the triangle with a drop input first, and then vary the n_periods to obtain different volume averages (so that it doesn't have to re-fit the triangle every time with all the exclusions)?

Thanks!

**Comment by jbogaardt** [comment]
I hate to give out hacks, but its better than fighting with perfomance. We should look into the performance issues (and bug you raised) as soon as we can.  One way to achieve the described workflow is to just hack a new triangle that omits the periods you want dropped.

```python
import chainladder as cl

raa = cl.load_sample('raa')

tri_hack = cl.Development(drop=[('1982', 12), ('1989', 24)]).fit_transform(raa)
tri_hack = tri_hack * tri_hack.nan_triangle # nan_triangle is undocumented feature and may chang in future
ldf_hack = cl.Development(n_periods=3).fit(tri_hack).ldf_
```

**Reply by contenrico** [reply]
Hi @jbogaardt 

Thanks for your reply. Perhaps I'm doing something wrong, but when I print the two results as shown below, I would expected to see a difference in the 12-24 LDF, while the results I get are exactly the same (i.e. it looks as if the drop input in `tri_hack` is not really doing anything):

```python
import chainladder as cl

raa = cl.load_sample('raa')

tri_hack = cl.Development(drop=[('1988', 12), ('1987', 12)]).fit_transform(raa)

tri_hack = tri_hack * tri_hack.nan_triangle # nan_triangle is undocumented feature and may chang in future

ldf_hack = cl.Development(n_periods=6).fit(tri_hack).ldf_

print(ldf_hack)
print(cl.Development(n_periods=6).fit(raa).ldf_)
```

**Reply by jbogaardt** [reply]
Sorry about that, I should have tested my recommendation more before sharing. You're correct that this doesn't work. I conflated `nan_triangle` with another implementation detail.

This is an even uglier hack that should work.  
```python
import chainladder as cl

raa = cl.load_sample('raa')

tri_hack = cl.Development(drop=[('1988', 12), ('1987', 12)]).fit_transform(raa)

tri_hack = tri_hack.iloc[..., :-1] * tri_hack.w_ + tri_hack.iloc[..., -1:] 

ldf_hack = cl.Development(n_periods=6).fit(tri_hack).ldf_

print(ldf_hack)
print(cl.Development(n_periods=6).fit(raa).ldf_)
```
I wouldn't want users to have to do this, this really should be improved in the package internals.

**Reply by contenrico** [reply]
Thanks @jbogaardt, that seems to work as a temporary fix - looking forward to future updates!

Thanks for all your work on the package.
